### ISSUE TITLE
Validate without variables fails when there is a subfolder present with TF code

### ISSUE THREAD

**Issue by 5obol** [antonbabenko/pre-commit-terraform#59]
Hi
My modules contains examples/ folder with implementation example. After upgrade of pre-commit-terraform and my code to 0.12.x the validate without variables stopped working for subfolders complaining that module wasn't initialised but it actually was.

My hook config:
```
- repo: git://github.com/antonbabenko/pre-commit-terraform
  rev: v1.17.0
  hooks:
    - id: terraform_fmt
    - id: terraform_docs
    - id: terraform_validate

Error:
pre-commit run -a
Terraform fmt............................................................Passed
Terraform docs...........................................................Passed
Terraform validate without variables.....................................Failed
hookid: terraform_validate

Success! The configuration is valid.


Failed path: examples
================================

Error: Module not installed

  on examples/main.tf line 6:
   6: module "s3_bucket" {

This module is not yet installed. Run "terraform init" to install all modules
required by this configuration.
```

**Comment by antonbabenko** [comment]
Hi Przemyslaw!

Thanks for opening this issue. Yes, `terraform validate` has slightly changed behavior in 0.12.

```
$ terraform validate -help

...

  Validation requires an initialized working directory with any referenced
  plugins and modules installed. To initialize a working directory for
  validation without accessing any configured remote backend, use:
      terraform init -backend=false

  To verify configuration in the context of a particular run (a particular
  target workspace, operation variables, etc), use the terraform plan
  subcommand instead, which includes an implied validation check.
```

I have not tried to reproduce the issue you are talking about, and I am not sure what we can do with this in this project (and whether we should?).

Please share your ideas.

**Comment by SolidSly** [comment]
Greetings,

Sorry to "reopen" this issue after two months. But I ran into kind of same problem today.
I will try to be as explicit as possible so everyone can understand how to reproduce it.

## Schema
.
├── parent_module_1
│   │
│   ├── child_module
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   └── variables.tf
│   │
│   ├── main.tf
│   ├── outputs.tf
│   └── variables.tf
│
├── parent_module_2
│   │
│   ├──child_module
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   └── variables.tf
│   │ 
│   ├── main.tf
│   ├── outputs.tf
│   └── variables.tf
│
├── main.tf
├── outputs.tf
└── variables.tf

## Problem

If you have a module inside another module and that the modules are all defined with a source type [local_path](https://www.terraform.io/docs/modules/sources.html#local-paths); then the pre-commit terraform validate will not work and return errors.

It is kind of logical since in the Terraform docs, they said the[local_path](https://www.terraform.io/docs/modules/sources.html#local-paths) modules are not really installed.

## Errors

```
Error: Module not installed

  on main.tf line 32:
  32: module "test" {

This module's local cache directory  could not be read. Run "terraform init"
to install all modules required by this configuration.
```

**Comment by kdimiche** [comment]
This may be a known issue with Terraform. See https://github.com/hashicorp/terraform/issues/21769.  Specifically:

> With all of this said, we recommend not relying on the directory arguments to these commands in general, because in any configuration that uses path.module and/or path.root it can cause details about the specific system running Terraform to be exposed into the state, making the configuration unportable to run on different systems. By far the most common and recommended way is to cd into the configuration directory and run all commands from there without any overriding path arguments.

Manually, I validated that running `terraform validate` works after changing directories:

```shell script
[AutoScaling]$ terraform validate
Success! The configuration is valid.

[AutoScaling]$ echo $?
0
[AutoScaling]$ cd test/
[test]$ terraform validate
Success! The configuration is valid.

[test]$ echo $?
0
```

**Comment by jmcshane** [comment]
So, reading into hashicorp/terraform#21769, running `terraform validate` against a subdirectory or module is just not something that is supported. I have to remove this check from my pre-commit configuration for now given the repository structure that I need to support.

Does terraform validate from the root validate all referenced files from submodules? At least variable references, I believe it does.

**Comment by antonbabenko** [comment]
`terraform_validate` hook has been just updated in #100, and [v1.29.0](https://github.com/antonbabenko/pre-commit-terraform/releases/v1.29.0) is now released. Please give it a try!

**Comment by antonbabenko** [comment]
Closing this issue.

I have released an updated version ([v1.44.0](https://github.com/antonbabenko/pre-commit-terraform/releases/tag/v1.44.0) ) where `terraform_validate` runs `terraform init` if no `.terraform` folder found.

See #158 and updated `README.md` file.

https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/68 - here you can see how I use both `terraform_validate` and `terraform_tflint`.
